server/hotrod/src/main/scala/org/infinispan/server/hotrod/iteration/IterationManager.scala

```diff
@@ -0,0 +1,132 @@
+package org.infinispan.server.hotrod.iteration
+
+import java.util
+import java.util.UUID
+
+import org.infinispan.commons.marshall.Marshaller
+import org.infinispan.commons.marshall.jboss.GenericJBossMarshaller
+import org.infinispan.commons.util.{CloseableIterator, CollectionFactory}
+import org.infinispan.container.entries.CacheEntry
+import org.infinispan.distribution.DistributionManager
+import org.infinispan.filter.KeyValueFilterConverter
+import org.infinispan.iteration.impl.EntryRetriever
+import org.infinispan.iteration.impl.EntryRetriever.SegmentListener
+import org.infinispan.manager.EmbeddedCacheManager
+import org.infinispan.server.hotrod.OperationStatus.OperationStatus
+import org.infinispan.server.hotrod._
+import org.infinispan.server.hotrod.logging.Log
+
+/**
+ * @author gustavonalle
+ * @since 8.0
+ */
+trait IterationManager {
+   type IterationId = String
+   def setFilterMarshallerClass(value: Class[_ <: Marshaller]): Unit
+   def startIteration(cacheName: String, segments: Bytes, filterConverterFactory: String, batch: Integer): IterationId
+   def closeIteration(cacheName: String, iterationId: IterationId): Boolean
+   def next(cacheName: String, iterationId: IterationId): IterationResult
+   def addKeyValueFilterConverter[K, V, C](factoryName: String, KeyValueFilterConverter: KeyValueFilterConverter[K, V, C])
+   def activeIterations: Int
+}
+
+class IterationSegmentsListener extends SegmentListener {
+   var finished = Set[Integer]()
+
+   def clear() = finished = finished.empty
+
+   override def segmentTransferred(segment: Int, sentLastEntry: Boolean): Unit = finished = finished + segment
+}
+
+class IterationState(val listener: IterationSegmentsListener, val iterator: CloseableIterator[CacheEntry[_, _]], val batch: Integer)
+
+object IterationState {
+   def apply(listener: IterationSegmentsListener, iterator: CloseableIterator[CacheEntry[_, _]], batch: Integer) = new IterationState(listener, iterator, batch)
+}
+
+sealed trait IterationResult {
```

Comment:
I like the use of sealed traits, but I'd try to hide implementations classes a bit unless other classes need something about the implementation itself. For example, I'd create an `IterationResult` object, and inside define the sealed trait impls as private classes, then have different factory (`apply`) methods for constructing either an empty one, with `apply()`...etc. By doing that, you'd encapsulate things.